README.md
# fiveN1 Rent Scraper

![logo](./logo/fiveN1-rent-scraper-logo.png)

![CI](https://github.com/neighborhood999/fiveN1-rent-scraper/workflows/CI/badge.svg?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/neighborhood999/fiveN1-rent-scraper)](https://goreportcard.com/report/github.com/neighborhood999/fiveN1-rent-scraper)
[![GoDoc](https://pkg.go.dev/badge/github.com/neighborhood999/fiveN1-rent-scraper?utm_source=godoc)](https://godoc.org/github.com/neighborhood999/fiveN1-rent-scraper)

> a.k.a 591 rent scraper.

Easy scraping [591](https://rent.591.com.tw/) rent information.

## Installation

```shell
$ make install
```

## Usage

```go
package main

import (
	"log"

	rent "github.com/neighborhood999/fiveN1-rent-scraper"
)

func main() {
	options := rent.NewOptions()
	url, err := rent.GenerateURL(options)
	if err != nil {
		log.Fatalf("\x1b[91;1m%s\x1b[0m", err)
	}

	f := rent.NewFiveN1(url)
	if err := f.Scrape(1); err != nil {
		log.Fatal(err)
	}

	json := rent.ConvertToJSON(f.RentList)
	log.Println(string(json))
}
```

And output json:

```json
{
  "1": [
    {
      "preview":
        "https://hp1.591.com.tw/house/active/2013/11/04/138356961541004002_765x517.water3.jpg",
      "title": "松山路套房 * 交通方便近捷運後山埤站 *",
      "url": "https://rent.591.com.tw/rent-detail-5857738.html",
      "address": "信義區 - 松山路 119 巷",
      "rentType": "沒有格局說明",
      "optionType": "獨立套房",
      "ping": "6",
      "floor": "樓層：3/4",
      "price": "9,000 元 / 月",
      "isNew": true
    }
  ]
}
```

Index number is the representation **page number**, every index contain **30** items. 🏠

## Options

Create and generate URL by options:

```go
options := rent.NewOptions()
url, err := rent.GenerateURL(options)
if err != nil {
	log.Fatal(err)
}

log.Println(url)
```

You may set more options for your requirement. Reference below `Options` struct:

```go
type Options struct {
	Region      int    `url:"region"`                // 地區 - 預設：`1`
	Section     string `url:"section,omitempty"`     // 鄉鎮 - 可選擇多個區域，例如：`section=7,4`
	Kind        int    `url:"kind"`                  // 租屋類型 - `0`：不限、`1`：整層住家、`2`：獨立套房、`3`：分租套房、`4`：雅房、`8`：車位，`24`：其他
	RentPrice   string `url:"rentprice,omitempty"`   // 租金 - `2`：5k - 10k、`3`：10k - 20k、`4`: 20k - 30k；或者可以輸入價格範圍，例如：`0,10000`
	Area        string `url:"area,omitempty"`        // 坪數格式 - `10,20`（10 到 20 坪）
	Order       string `url:"order"`                 // 貼文時間 - 預設使用刊登時間：`posttime`，或是使用價格排序：`money`
	OrderType   string `url:"orderType"`             // 排序方式 - `desc` 或 `asc`
	Sex         int    `url:"sex,omitempty"`         // 性別 - `0`：不限、`1`：男性、`2`：女性
	HasImg      string `url:"hasimg,omitempty"`      // 過濾是否有「房屋照片」 - ``：空值（不限）、`1`：是
	NotCover    string `url:"not_cover,omitempty"`   // 過濾是否為「頂樓加蓋」 - ``：空值（不限）、`1`：是
	Role        string `url:"role,omitempty"`        // 過濾是否為「屋主刊登」 - ``：空值（不限）、`1`：是
	Shape       string `url:"shape,omitempty"`       // 房屋類型 - `1`：公寓、`2`：電梯大樓、`3`：透天厝、`4`：別墅
	Pattern     string `url:"pattern,omitempty"`     // 格局單選 - `0`：不限、`1`：一房、`2``：兩房、`3`：三房、`4`：四房、`5`：五房以上
	PatternMore string `url:"patternMore,omitempty"` // 格局多選 - 參考「格局單選」，可以選多種格局，例如：`1,2,3,4,5`
	Floor       string `url:"floor,omitempty"`       // 樓層 - `0,0`：不限、`0,1`：一樓、`2,6`：二樓到六樓、`6,12`：六樓到十二樓、`12,`：十二樓以上
	Option      string `url:"option,omitempty"`      // 提供設備 - `tv`：電視、`cold`：冷氣、`icebox`：冰箱、`hotwater`：熱水器、`naturalgas`：天然瓦斯、`four`：第四台、`broadband`：網路、`washer`：洗衣機、`bed`：床、`wardrobe`：衣櫃、`sofa`：沙發。可選擇多個設備，例如：option=tv,cold
	Other       string `url:"other,omitempty"`       // 其他條件 - `cartplace`：有車位、`lift`：有電梯、`balcony_1`：有陽台、`cook`：可開伙、`pet`：可養寵物、`tragoods`：近捷運、`lease`：可短期租賃。可選擇多個條件，例如：other=cartplace,cook
	FirstRow    int    `url:"firstRow"`
}
```

For example:

```go
options := rent.NewOptions()
options.Kind = 3
options.HasImg = "1"
options.NotCover = "1"
options.Role = "1"

rent.GenerateURL(options)
```

## Multiple Pages

If you want to get more rent pieces of information, setting page amount in `Scrape` method:

```go
f := rent.NewFiveN1(url)
f.Scrape(5) // page.1 to page.5
```

When amount `> 1`, it will start goroutine automatically and correspond to the page number to scrape.

## Code Lists

- [地區列表](./list/url-jump-ip.md)
- [鄉鎮列表](./list/section.md)

## LICENSE

MIT © [Peng Jie](https://github.com/neighborhood999)


CHANGELOG.md
# Change Log

# [1.2.4](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/v1.2.3...v1.2.4)(2021-09-22)
### Fixes
- fixed: remove `https` prefix (#51)

# [1.2.3](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/v1.2.2...v1.2.3)(2021-05-07)
### Chores
- chores: export `TotalPages` vairable (#50)

# [1.2.2](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/v1.2.1...v1.2.2)(2021-04-26)
### Refactor
- refactor: add `isExecuted` flag to avoid re-fetch (#49)

# [1.2.0](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/1.1.1...1.2.0)(2018-07-23)
### Features
- 新增 `pattern` 和 `patternMore` **格局**選項
- **各縣市地區代號** 移至 [list](https://github.com/neighborhood999/fiveN1-rent-scraper/tree/master/list)

# [1.1.1](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/1.1.0...1.1.1)(2018-05-22)
### Fixes
- 修正 `orderType` 參數 ([#36](https://github.com/neighborhood999/fiveN1-rent-scraper/pull/36))

# [1.1.0](https://github.com/neighborhood999/fiveN1-rent-scraper/compare/1.0.0...1.1.0) (2018-05-17)
### Features
- 新增進階搜尋功能的參數 ([#34](https://github.com/neighborhood999/fiveN1-rent-scraper/pull/34))

# 1.0.0 (2018-05-16)
### Features
- 可自定義搜尋參數
- 支援輸出租屋資訊 JSON 格式資料
- 併發（Concurrency）爬取多頁租屋資訊 ([#5](https://github.com/neighborhood999/fiveN1-rent-scraper/pull/5))

- 完整城市、鄉鎮參數列表 ([#13](https://github.com/neighborhood999/fiveN1-rent-scraper/pull/13), [#15](https://github.com/neighborhood999/fiveN1-rent-scraper/pull/15))


rent_test.go
package rent

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
)

const (
	noMorePageErrMessage = "\x1b[91;1mNo More Pages！\x1b[0m"
)

func callbackHandler(w http.ResponseWriter, r *http.Request) {
	html, _ := ioutil.ReadFile("_fixtures/index.html")
	w.Header().Set("Content-Type", "application/html")
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(html))
}

func TestGenerateURL(t *testing.T) {
	var err error
	o := NewOptions()

	_, err = GenerateURL(o)
	assert.EqualValues(t, nil, err)
}

func TestSetReqCookie(t *testing.T) {
	o := NewOptions()
	url, _ := GenerateURL(o)
	f := NewFiveN1(url)
	f.SetReqCookie("2")

	assert.Equal(t, "urlJumpIp=2", f.cookie.String())
}

func TestConvertToJSON(t *testing.T) {
	hm := make(map[int][]*HouseInfo)
	h := NewHouseInfo()

	h.Title = "Test Rent House Inforation"
	h.URL = "https://github.com/neighborhood999/fiveN1-rent-scraper"
	h.Address = "Taipei, Taiwan"
	h.RentType = "沒有格局說明"
	h.OptionType = "獨立套房"
	h.Ping = "9"
	h.Floor = "樓層：7/9"
	h.Price = "10,000 元 / 月"
	h.IsNew = true

	s := make([]*HouseInfo, 1)
	s[0] = h
	hm[0] = s

	b := ConvertToJSON(hm)

	var expectedType []byte
	assert.IsType(t, expectedType, b)
}

func TestExportJSON(t *testing.T) {
	text := []byte("Hello, World")
	ExportJSON(text)
}

func TestScrape(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(callbackHandler))
	defer server.Close()
	mockURL := server.URL

	expectedPages := 12
	expectedRecords := 333

	f := NewFiveN1(mockURL)
	f.Scrape(1)

	assert.Equal(t, expectedPages, f.TotalPages)
	assert.Equal(t, expectedRecords, f.records)

	err := f.Scrape(13)
	assert.EqualError(t, err, noMorePageErrMessage)
}

func TestConcurrentScrape(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(callbackHandler))
	defer server.Close()
	mockURL := server.URL + "/?"

	f := NewFiveN1(mockURL)
	f.Scrape(2)

	assert.IsType(t, HouseInfoCollection{}, f.RentList)
	assert.Equal(t, 2, len(f.RentList))
}

func TestGetTotalPage(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(callbackHandler))
	defer server.Close()
	mockURL := server.URL

	f := NewFiveN1(mockURL)
	assert.Equal(t, 0, f.TotalPages)
	assert.Equal(t, false, f.isExecuted)

	f.GetTotalPage()
	assert.Equal(t, 12, f.TotalPages)
	assert.Equal(t, true, f.isExecuted)
}


_example/api-server/server.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"

	rent "github.com/neighborhood999/fiveN1-rent-scraper"
)

type RentInformation struct {
	TotalPages int               `json:"totalPages"`
	Data       []*rent.HouseInfo `json:"data"`
}

func callbackHandler(w http.ResponseWriter, r *http.Request) {
	options := rent.NewOptions()
	r.ParseForm()

	if r.Method == "GET" {
		options.Area = r.FormValue("area")
		options.Role = r.FormValue("role")
		options.Other = r.FormValue("other")
		options.Order = r.FormValue("order")
		options.Shape = r.FormValue("shape")
		options.Floor = r.FormValue("floor")
		options.Option = r.FormValue("option")
		options.HasImg = r.FormValue("hasImage")
		options.Section = r.FormValue("section")
		options.NotCover = r.FormValue("notCover")
		options.RentPrice = r.FormValue("rentPrice")
		options.OrderType = r.FormValue("orderType")
		options.Sex, _ = strconv.Atoi(r.FormValue("sex"))
		options.Kind, _ = strconv.Atoi(r.FormValue("kind"))
		options.FirstRow, _ = strconv.Atoi(r.FormValue("firstRow"))

		url, err := rent.GenerateURL(options)
		if err != nil {
			log.Fatal(err)
		}

		f := rent.NewFiveN1(url)
		f.SetReqCookie(r.FormValue("urlJump"))

		if err := f.Scrape(1); err != nil {
			log.Fatal(err)
		}

		response, err := json.Marshal(RentInformation{
			TotalPages: f.TotalPages,
			Data:       f.RentList[1],
		})

		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Credentials", "true")
		w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT, DELETE")
		w.Header().Set("Access-Control-Allow-Headers", "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token")

		w.Write(response)
	}
}

func main() {
	http.HandleFunc("/", callbackHandler)
	http.ListenAndServe(":8080", nil)
}


rent.go
package rent

import (
	"encoding/json"
	"errors"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/google/go-querystring/query"
	"github.com/vinta/pangu"
)

const (
	rootURL = "https://rent.591.com.tw/"
)

// HouseInfo is the representation rent house information.
type HouseInfo struct {
	Preview    string `json:"preview"`
	Title      string `json:"title"`
	URL        string `json:"url"`
	Address    string `json:"address"`
	RentType   string `json:"rentType"`
	OptionType string `json:"optionType"`
	Ping       string `json:"ping"` // a.k.a 坪數
	Floor      string `json:"floor"`
	Price      string `json:"price"`
	IsNew      bool   `json:"isNew"`
}

// Options is the representation query argument.
type Options struct {
	Region      int    `url:"region"`                // 地區 - 預設：`1`
	Section     string `url:"section,omitempty"`     // 鄉鎮 - 可選擇多個區域，例如：`section=7,4`
	Kind        int    `url:"kind"`                  // 租屋類型 - `0`：不限、`1`：整層住家、`2`：獨立套房、`3`：分租套房、`4`：雅房、`8`：車位，`24`：其他
	RentPrice   string `url:"rentprice,omitempty"`   // 租金 - `2`：5k - 10k、`3`：10k - 20k、`4`: 20k - 30k；或者可以輸入價格範圍，例如：`0,10000`
	Area        string `url:"area,omitempty"`        // 坪數格式 - `10,20`（10 到 20 坪）
	Order       string `url:"order"`                 // 貼文時間 - 預設使用刊登時間：`posttime`，或是使用價格排序：`money`
	OrderType   string `url:"orderType"`             // 排序方式 - `desc` 或 `asc`
	Sex         int    `url:"sex,omitempty"`         // 性別 - `0`：不限、`1`：男性、`2`：女性
	HasImg      string `url:"hasimg,omitempty"`      // 過濾是否有「房屋照片」 - ``：空值（不限）、`1`：是
	NotCover    string `url:"not_cover,omitempty"`   // 過濾是否為「頂樓加蓋」 - ``：空值（不限）、`1`：是
	Role        string `url:"role,omitempty"`        // 過濾是否為「屋主刊登」 - ``：空值（不限）、`1`：是
	Shape       string `url:"shape,omitempty"`       // 房屋類型 - `1`：公寓、`2`：電梯大樓、`3`：透天厝、`4`：別墅
	Pattern     string `url:"pattern,omitempty"`     // 格局單選 - `0`：不限、`1`：一房、`2``：兩房、`3`：三房、`4`：四房、`5`：五房以上
	PatternMore string `url:"patternMore,omitempty"` // 格局多選 - 參考「格局單選」，可以選多種格局，例如：`1,2,3,4,5`
	Floor       string `url:"floor,omitempty"`       // 樓層 - `0,0`：不限、`0,1`：一樓、`2,6`：二樓到六樓、`6,12`：六樓到十二樓、`12,`：十二樓以上
	Option      string `url:"option,omitempty"`      // 提供設備 - `tv`：電視、`cold`：冷氣、`icebox`：冰箱、`hotwater`：熱水器、`naturalgas`：天然瓦斯、`four`：第四台、`broadband`：網路、`washer`：洗衣機、`bed`：床、`wardrobe`：衣櫃、`sofa`：沙發。可選擇多個設備，例如：option=tv,cold
	Other       string `url:"other,omitempty"`       // 其他條件 - `cartplace`：有車位、`lift`：有電梯、`balcony_1`：有陽台、`cook`：可開伙、`pet`：可養寵物、`tragoods`：近捷運、`lease`：可短期租賃。可選擇多個條件，例如：other=cartplace,cook
	FirstRow    int    `url:"firstRow"`
}

// HouseInfoCollection is the representation house information collection.
type HouseInfoCollection map[int][]*HouseInfo

// Document is the representation goquery.Document.
type Document struct {
	doc *goquery.Document
}

// FiveN1 is the representation page information.
type FiveN1 struct {
	records    int
	TotalPages int
	queryURL   string
	isExecuted bool
	RentList   HouseInfoCollection
	wg         sync.WaitGroup
	rw         sync.RWMutex
	client     *http.Client
	cookie     *http.Cookie
}

// NewHouseInfo create a new `HouseInfo`.
func NewHouseInfo() *HouseInfo {
	return &HouseInfo{}
}

// NewOptions create a `Options` with default value.
func NewOptions() *Options {
	return &Options{
		Kind:      2,
		Region:    1,
		Section:   "0",
		RentPrice: "2",
		Order:     "posttime",
		OrderType: "desc",
		FirstRow:  0,
	}
}

// NewFiveN1 create a `FiveN1` with default value.
func NewFiveN1(url string) *FiveN1 {
	return &FiveN1{
		queryURL: url,
		RentList: make(map[int][]*HouseInfo),
		client:   &http.Client{},
		cookie: &http.Cookie{
			Name:  "urlJumpIp",
			Value: "1",
		},
	}
}

// NewDocument create a `Document` with default value.
func NewDocument() *Document {
	return &Document{
		doc: &goquery.Document{},
	}
}

// GenerateURL is convert options to query parameters.
func GenerateURL(o *Options) (string, error) {
	v, error := query.Values(o)

	return rootURL + "?" + v.Encode(), error
}

func stringReplacer(text string) string {
	replacer := strings.NewReplacer("\n", "", " ", "")

	return pangu.SpacingText(replacer.Replace(text))
}

func trimTextSpace(s string) string {
	return strings.Fields(s)[0]
}

func fillDescription(s []string) []string {
	s = append(s, s[2])
	s[2] = "沒有格局說明"

	return s
}

// ConvertToJSON is convert rent house info collection to json.
func ConvertToJSON(list HouseInfoCollection) []byte {
	b, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		log.Fatal(err)
	}

	return b
}

// ExportJSON export json file.
func ExportJSON(b []byte) {
	f, err := os.Create("/tmp/rent.json")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	f.Write(b)
	log.Println("✔️  Export Path: \x1b[42m/tmp/rent.json\x1b[0m")
}

func (d *Document) clone(res *http.Response) {
	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		log.Fatal(err)
	}

	d.doc = doc
}

// SetReqCookie set the region value.
func (f *FiveN1) SetReqCookie(region string) {
	f.cookie.Value = region
}

func (f *FiveN1) request(url string) *http.Response {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		log.Fatal(err)
	}

	req.AddCookie(f.cookie)

	res, err := f.client.Do(req)
	if err != nil {
		log.Fatal(err)
	}

	return res
}

func (f *FiveN1) parseRentHouse(page int, doc *goquery.Document) {
	doc.Find("#content").Each(func(_ int, selector *goquery.Selection) {
		selector.Find(".listInfo.clearfix").Each(func(item int, listInfo *goquery.Selection) {
			houseInfo := NewHouseInfo()

			// Content Title
			title := listInfo.Find(".pull-left.infoContent > h3 > a[href]").Text()
			houseInfo.Title = stringReplacer(title)

			// Content URL
			var url string
			if href, ok := listInfo.Find(".pull-left.infoContent > h3 > a").Attr("href"); ok {
				url = stringReplacer(href)
			}
			houseInfo.URL = url

			if crop, ok := listInfo.Find(".pull-left.imageBox > img").Attr("data-original"); ok {
				preview := strings.Replace(crop, "210x158.crop.jpg", "765x517.water3.jpg", 1)
				houseInfo.Preview = preview
			}

			listInfo.Find(".pull-left.infoContent").Each(func(_ int, infoContent *goquery.Selection) {
				// Rent House Description.
				description := stringReplacer(infoContent.Find(".lightBox").First().Text())

				splitDescription := strings.Split(description, "|")

				// Exchange
				if len(splitDescription) == 4 {
					tmp := splitDescription[2] // 坪數
					splitDescription[2] = splitDescription[1]
					splitDescription[1] = tmp
				}

				if len(splitDescription) < 4 {
					splitDescription = fillDescription(splitDescription)
				}

				houseInfo.OptionType = trimTextSpace(splitDescription[0])
				houseInfo.Ping = trimTextSpace(splitDescription[1])
				houseInfo.RentType = trimTextSpace(splitDescription[2])
				houseInfo.Floor = trimTextSpace(splitDescription[3])

				// Rent House Address
				address := stringReplacer(infoContent.Find(".lightBox").Eq(1).Text())
				houseInfo.Address = address
			})

			// Rent Price
			listInfo.Find(".price").Each(func(_ int, price *goquery.Selection) {
				houseInfo.Price = stringReplacer(price.Text())
			})

			// New Rent House
			listInfo.Find(".newArticle").Each(func(_ int, n *goquery.Selection) {
				houseInfo.IsNew = true
			})

			// Add rent house into list
			f.rw.Lock()
			f.RentList[page+1] = append(f.RentList[page+1], houseInfo)
			f.rw.Unlock()
		})
	})
}

func (f *FiveN1) parseRecordsNum(doc *goquery.Document) {
	doc.Find(".pull-left.hasData > i").Each(func(_ int, selector *goquery.Selection) {
		recordString := stringReplacer(selector.Text())
		replaceComman := strings.Replace(recordString, ",", "", -1)
		totalRecord, _ := strconv.Atoi(replaceComman)
		f.records = totalRecord
		f.TotalPages = (totalRecord / 30) + 1
	})
}

func (f *FiveN1) firstPage() {
	d := NewDocument()
	res := f.request(f.queryURL)
	d.clone(res)

	f.isExecuted = true
	f.parseRecordsNum(d.doc) // Record pages number at first
	log.Println("---------------------")
	log.Printf("| Query URL:    \x1b[94;1m%v\x1b[0m |", f.queryURL)
	log.Printf("| Total Record: \x1b[94;1m%d\x1b[0m |", f.records)
	log.Printf("| Total Page:   \x1b[94;1m%d\x1b[0m  |", f.TotalPages)
	log.Println("---------------------")
	f.parseRentHouse(0, d.doc)
}

func (f *FiveN1) worker(n int) {
	defer f.wg.Done()
	log.Printf("\x1b[30;1mStart worker at page number: %d\x1b[0m", n+1)

	d := NewDocument()
	r := strconv.Itoa(n * 30)
	res := f.request(f.queryURL + "&firstRow=" + r)
	d.clone(res)

	f.parseRentHouse(n, d.doc)
}

// Scrape will clone entire DOM for reuse.
// It will scrape first page at first, if specify page number > 1,
// it will start workers.
func (f *FiveN1) Scrape(page int) error {
	f.firstPage()
	if page > f.TotalPages {
		return errors.New("\x1b[91;1mNo More Pages！\x1b[0m")
	}

	for i := 1; i < page; i++ {
		f.wg.Add(1)
		go f.worker(i)
	}

	f.wg.Wait()

	return nil
}

// GetTotalPage can use get total page
func (f *FiveN1) GetTotalPage() int {
	if !f.isExecuted {
		f.firstPage()
	}

	return f.TotalPages
}
